<a href="https://colab.research.google.com/github/grupo21dex/deliverable-1/blob/main/Din%C3%A2mica_NYFlights_Grupo21.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd


In [3]:
df = pd.read_csv("nyflights.csv")

In [4]:
df.head(5)

Unnamed: 0  year  month  day  dep_time  sched_dep_time  dep_delay  \
0           1  2013      1    1     517.0             515        2.0   
1           2  2013      1    1     533.0             529        4.0   
2           3  2013      1    1     542.0             540        2.0   
3           4  2013      1    1     544.0             545       -1.0   
4           5  2013      1    1     554.0             600       -6.0   

   arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0     830.0             819       11.0      UA    1545  N14228    EWR  IAH   
1     850.0             830       20.0      UA    1714  N24211    LGA  IAH   
2     923.0             850       33.0      AA    1141  N619AA    JFK  MIA   
3    1004.0            1022      -18.0      B6     725  N804JB    JFK  BQN   
4     812.0             837      -25.0      DL     461  N668DN    LGA  ATL   

   air_time  distance  hour  minute            time_hour  
0     227.0      1400     5      15  2013-01-01 05:00:00  
1     227.0      1416     5      29  2013-01-01 05:00:00  
2     160.0      1089     5      40  2013-01-01 05:00:00  
3     183.0      1576     5      45  2013-01-01 05:00:00  
4     116.0       762     6       0  2013-01-01 06:00:00

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19558 entries, 0 to 19557
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      19558 non-null  int64  
 1   year            19558 non-null  int64  
 2   month           19558 non-null  int64  
 3   day             19558 non-null  int64  
 4   dep_time        19380 non-null  float64
 5   sched_dep_time  19558 non-null  int64  
 6   dep_delay       19380 non-null  float64
 7   arr_time        19371 non-null  float64
 8   sched_arr_time  19558 non-null  int64  
 9   arr_delay       19325 non-null  float64
 10  carrier         19558 non-null  object 
 11  flight          19558 non-null  int64  
 12  tailnum         19491 non-null  object 
 13  origin          19558 non-null  object 
 14  dest            19558 non-null  object 
 15  air_time        19325 non-null  float64
 16  distance        19558 non-null  int64  
 17  hour            19558 non-null 

In [15]:
#Etapa 1

df_filtro = df[df["dep_delay"]>120]
df_filtro.head()

Unnamed: 0  year  month  day  dep_time  sched_dep_time  dep_delay  \
151         152  2013      1    1     848.0            1835      853.0   
218         219  2013      1    1     957.0             733      144.0   
268         269  2013      1    1    1114.0             900      134.0   
491         492  2013      1    1    1540.0            1338      122.0   
649         650  2013      1    1    1815.0            1325      290.0   

     arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
151    1001.0            1950      851.0      MQ    3944  N942MQ    JFK  BWI   
218    1056.0             853      123.0      UA     856  N534UA    EWR  BOS   
268    1447.0            1222      145.0      UA    1086  N76502    LGA  IAH   
491    2020.0            1825      115.0      B6     705  N570JB    JFK  SJU   
649    2120.0            1542      338.0      EV    4417  N17185    EWR  OMA   

     air_time  distance  hour  minute            time_hour  
151      41.0       184    18      35  2013-01-01 18:00:00  
218      37.0       200     7      33  2013-01-01 07:00:00  
268     248.0      1416     9       0  2013-01-01 09:00:00  
491     193.0      1598    13      38  2013-01-01 13:00:00  
649     213.0      1134    13      25  2013-01-01 13:00:00

In [16]:
df_describe = df_filtro["dep_delay"].describe().T.copy()
df_describe

count     296.000000
mean      188.300676
std       113.154675
min       121.000000
25%       135.000000
50%       157.000000
75%       198.250000
max      1301.000000
Name: dep_delay, dtype: float64

In [17]:
df_describe = df_filtro.describe().T.copy()
df_describe

count          mean          std     min      25%      50%  \
Unnamed: 0      296.0  10881.351351  6096.514584   152.0  4633.00  11849.0   
year            296.0   2013.000000     0.000000  2013.0  2013.00   2013.0   
month           296.0      1.000000     0.000000     1.0     1.00      1.0   
day             296.0     12.780405     6.998965     1.0     6.00     14.0   
dep_time        296.0   1679.192568   552.710308     2.0  1306.50   1837.5   
sched_dep_time  296.0   1497.783784   450.935758   558.0  1126.75   1619.0   
dep_delay       296.0    188.300676   113.154675   121.0   135.00    157.0   
arr_time        295.0   1426.000000   819.473043     2.0  1001.50   1645.0   
sched_arr_time  296.0   1687.510135   484.835425    36.0  1332.00   1820.0   
arr_delay       293.0    184.965870   114.190115    65.0   133.00    160.0   
flight          296.0   2745.057432  1823.855635     7.0   622.25   3490.0   
air_time        293.0    138.372014    89.965641    26.0    69.00    124.0   
distance        296.0    896.331081   707.492544    80.0   397.00    736.5   
hour            296.0     14.722973     4.493794     5.0    11.00     16.0   
minute          296.0     25.486486    18.538640     0.0     8.00     29.0   

                     75%      max  
Unnamed: 0      15663.25  19508.0  
year             2013.00   2013.0  
month               1.00      1.0  
day                18.00     23.0  
dep_time         2115.25   2359.0  
sched_dep_time   1851.25   2359.0  
dep_delay         198.25   1301.0  
arr_time         2127.50   2358.0  
sched_arr_time   2055.25   2359.0  
arr_delay         199.00   1272.0  
flight           4317.25   6055.0  
air_time          172.00    640.0  
distance         1092.00   4983.0  
hour               18.00     23.0  
minute             40.00     59.0

In [18]:
df_describe["amplitude"] = df_describe['max'] - df_describe['min']
df_describe["variância"] = df_describe['std'] **2
df_describe

count          mean          std     min      25%      50%  \
Unnamed: 0      296.0  10881.351351  6096.514584   152.0  4633.00  11849.0   
year            296.0   2013.000000     0.000000  2013.0  2013.00   2013.0   
month           296.0      1.000000     0.000000     1.0     1.00      1.0   
day             296.0     12.780405     6.998965     1.0     6.00     14.0   
dep_time        296.0   1679.192568   552.710308     2.0  1306.50   1837.5   
sched_dep_time  296.0   1497.783784   450.935758   558.0  1126.75   1619.0   
dep_delay       296.0    188.300676   113.154675   121.0   135.00    157.0   
arr_time        295.0   1426.000000   819.473043     2.0  1001.50   1645.0   
sched_arr_time  296.0   1687.510135   484.835425    36.0  1332.00   1820.0   
arr_delay       293.0    184.965870   114.190115    65.0   133.00    160.0   
flight          296.0   2745.057432  1823.855635     7.0   622.25   3490.0   
air_time        293.0    138.372014    89.965641    26.0    69.00    124.0   
distance        296.0    896.331081   707.492544    80.0   397.00    736.5   
hour            296.0     14.722973     4.493794     5.0    11.00     16.0   
minute          296.0     25.486486    18.538640     0.0     8.00     29.0   

                     75%      max  amplitude     variância  
Unnamed: 0      15663.25  19508.0    19356.0  3.716749e+07  
year             2013.00   2013.0        0.0  0.000000e+00  
month               1.00      1.0        0.0  0.000000e+00  
day                18.00     23.0       22.0  4.898551e+01  
dep_time         2115.25   2359.0     2357.0  3.054887e+05  
sched_dep_time   1851.25   2359.0     1801.0  2.033431e+05  
dep_delay         198.25   1301.0     1180.0  1.280398e+04  
arr_time         2127.50   2358.0     2356.0  6.715361e+05  
sched_arr_time   2055.25   2359.0     2323.0  2.350654e+05  
arr_delay         199.00   1272.0     1207.0  1.303938e+04  
flight           4317.25   6055.0     6048.0  3.326449e+06  
air_time          172.00    640.0      614.0  8.093817e+03  
distance         1092.00   4983.0     4903.0  5.005457e+05  
hour               18.00     23.0       18.0  2.019418e+01  
minute             40.00     59.0       59.0  3.436812e+02

O comportamento médio foi 188.300676
A variância foi 1.280398e+04

In [21]:
#Etapa 2
df_atraso = df[df["dep_delay"]>5]
df_atraso

Unnamed: 0  year  month  day  dep_time  sched_dep_time  dep_delay  \
25             26  2013      1    1     608.0             600        8.0   
26             27  2013      1    1     611.0             600       11.0   
31             32  2013      1    1     623.0             610       13.0   
41             42  2013      1    1     632.0             608       24.0   
47             48  2013      1    1     644.0             636        8.0   
...           ...   ...    ...  ...       ...             ...        ...   
19535       19536  2013      1   23    1324.0            1315        9.0   
19550       19551  2013      1   23    1351.0            1345        6.0   
19551       19552  2013      1   23    1352.0            1250       62.0   
19552       19553  2013      1   23    1352.0            1335       17.0   
19557       19558  2013      1   23    1357.0            1329       28.0   

       arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin  \
25        807.0             735       32.0      MQ    3768  N9EAMQ    EWR   
26        945.0             931       14.0      UA     303  N532UA    JFK   
31        920.0             915        5.0      AA    1837  N3EMAA    LGA   
41        740.0             728       12.0      EV    4144  N13553    EWR   
47        931.0             940       -9.0      UA    1701  N75435    EWR   
...         ...             ...        ...     ...     ...     ...    ...   
19535    1541.0            1520       21.0      MQ    4564  N730MQ    LGA   
19550    1637.0            1641       -4.0      B6    1783  N524JB    JFK   
19551    1452.0            1355       57.0      B6    1006  N229JB    JFK   
19552    1651.0            1633       18.0      B6    1161  N613JB    LGA   
19557    1605.0            1526       39.0      EV    3840  N13994    EWR   

      dest  air_time  distance  hour  minute            time_hour  
25     ORD     139.0       719     6       0  2013-01-01 06:00:00  
26     SFO     366.0      2586     6       0  2013-01-01 06:00:00  
31     MIA     153.0      1096     6      10  2013-01-01 06:00:00  
41     IAD      52.0       212     6       8  2013-01-01 06:00:00  
47     FLL     151.0      1065     6      36  2013-01-01 06:00:00  
...    ...       ...       ...   ...     ...                  ...  
19535  DTW      93.0       502    13      15  2013-01-23 13:00:00  
19550  MCO     145.0       944    13      45  2013-01-23 13:00:00  
19551  BOS      37.0       187    12      50  2013-01-23 12:00:00  
19552  PBI     152.0      1035    13      35  2013-01-23 13:00:00  
19557  MEM     154.0       946    13      29             2013-01-  

[4835 rows x 20 columns]

In [28]:
df_atraso["dest"].value_counts().head(10)

ORD    223
FLL    214
MCO    205
BOS    193
ATL    169
LAX    166
MIA    157
CLT    145
DCA    136
SFO    136
Name: dest, dtype: int64

In [27]:
df_atraso["dest"].value_counts().tail(10)

BHM    5
OAK    5
SJC    4
CAE    4
CAK    4
AVL    1
BZN    1
EYW    1
PSP    1
HDN    1
Name: dest, dtype: int64

Os 5 destinos que tiveram mais atrasos na partida são: 
ORD = 223;
FLL = 214;
MCO = 205;
BOS = 193;
ATL = 169

In [29]:
#Etapa 3

lista_destino = ['ORD', 'FLL', 'MCO', 'BOS', 'ATL']
df_med_piores = df_atraso[df_atraso.dest.isin(lista_destino)]
df_med_piores

Unnamed: 0  year  month  day  dep_time  sched_dep_time  dep_delay  \
25             26  2013      1    1     608.0             600        8.0   
47             48  2013      1    1     644.0             636        8.0   
70             71  2013      1    1     709.0             700        9.0   
85             86  2013      1    1     732.0             645       47.0   
96             97  2013      1    1     749.0             710       39.0   
...           ...   ...    ...  ...       ...             ...        ...   
19474       19475  2013      1   23    1156.0            1130       26.0   
19498       19499  2013      1   23    1242.0            1230       12.0   
19522       19523  2013      1   23    1312.0            1300       12.0   
19550       19551  2013      1   23    1351.0            1345        6.0   
19551       19552  2013      1   23    1352.0            1250       62.0   

       arr_time  sched_arr_time  arr_delay carrier  flight tailnum origin  \
25        807.0             735       32.0      MQ    3768  N9EAMQ    EWR   
47        931.0             940       -9.0      UA    1701  N75435    EWR   
70        852.0             832       20.0      UA    1092  N26226    LGA   
85       1011.0             941       30.0      UA    1111  N37456    EWR   
96        939.0             850       49.0      MQ    3737  N508MQ    EWR   
...         ...             ...        ...     ...     ...     ...    ...   
19474    1456.0            1430       26.0      DL     695  N687DL    JFK   
19498    1459.0            1459        0.0      EV    3826  N15985    EWR   
19522    1606.0            1606        0.0      B6      85  N589JB    JFK   
19550    1637.0            1641       -4.0      B6    1783  N524JB    JFK   
19551    1452.0            1355       57.0      B6    1006  N229JB    JFK   

      dest  air_time  distance  hour  minute            time_hour  
25     ORD     139.0       719     6       0  2013-01-01 06:00:00  
47     FLL     151.0      1065     6      36  2013-01-01 06:00:00  
70     ORD     135.0       733     7       0  2013-01-01 07:00:00  
85     MCO     145.0       937     6      45  2013-01-01 06:00:00  
96     ORD     148.0       719     7      10  2013-01-01 07:00:00  
...    ...       ...       ...   ...     ...                  ...  
19474  MCO     142.0       944    11      30  2013-01-23 11:00:00  
19498  ATL     121.0       746    12      30  2013-01-23 12:00:00  
19522  FLL     156.0      1069    13       0  2013-01-23 13:00:00  
19550  MCO     145.0       944    13      45  2013-01-23 13:00:00  
19551  BOS      37.0       187    12      50  2013-01-23 12:00:00  

[1004 rows x 20 columns]

In [30]:
df_med_piores.describe()

Unnamed: 0    year   month          day     dep_time  sched_dep_time  \
count   1004.000000  1004.0  1004.0  1004.000000  1004.000000     1004.000000   
mean    9764.350598  2013.0     1.0    11.600598  1498.165339     1448.915339   
std     5905.549549     0.0     0.0     6.787975   457.795962      446.609900   
min       26.000000  2013.0     1.0     1.000000    10.000000      558.000000   
25%     4093.250000  2013.0     1.0     5.000000  1154.500000     1123.000000   
50%    10332.000000  2013.0     1.0    12.000000  1557.000000     1529.000000   
75%    14782.250000  2013.0     1.0    17.000000  1830.250000     1800.000000   
max    19552.000000  2013.0     1.0    23.000000  2351.000000     2258.000000   

         dep_delay     arr_time  sched_arr_time    arr_delay       flight  \
count  1004.000000  1004.000000     1004.000000  1003.000000  1004.000000   
mean     36.779880  1600.618526     1614.373506    33.000997  1279.254980   
std      59.665939   548.695092      501.207198    61.311343  1227.257335   
min       6.000000     3.000000        2.000000   -28.000000     1.000000   
25%      11.000000  1247.500000     1253.500000     4.000000   379.250000   
50%      20.000000  1748.500000     1742.000000    19.000000   954.000000   
75%      40.000000  2015.250000     2005.000000    42.000000  1703.000000   
max    1126.000000  2359.000000     2359.000000  1109.000000  5621.000000   

          air_time     distance         hour       minute  
count  1003.000000  1004.000000  1004.000000  1004.000000  
mean    119.010967   746.715139    14.249004    24.014940  
std      41.664314   298.390492     4.445723    19.511501  
min      26.000000   184.000000     5.000000     0.000000  
25%     116.000000   719.000000    11.000000     5.000000  
50%     130.000000   762.000000    15.000000    21.000000  
75%     147.000000   950.000000    18.000000    40.000000  
max     188.000000  1076.000000    22.000000    59.000000

In [31]:
df_med_piores['velocidade média'] = df_med_piores["distance"]/df_med_piores["air_time"]
df_med_piores

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Unnamed: 0  year  month  day  dep_time  sched_dep_time  dep_delay  \
25             26  2013      1    1     608.0             600        8.0   
47             48  2013      1    1     644.0             636        8.0   
70             71  2013      1    1     709.0             700        9.0   
85             86  2013      1    1     732.0             645       47.0   
96             97  2013      1    1     749.0             710       39.0   
...           ...   ...    ...  ...       ...             ...        ...   
19474       19475  2013      1   23    1156.0            1130       26.0   
19498       19499  2013      1   23    1242.0            1230       12.0   
19522       19523  2013      1   23    1312.0            1300       12.0   
19550       19551  2013      1   23    1351.0            1345        6.0   
19551       19552  2013      1   23    1352.0            1250       62.0   

       arr_time  sched_arr_time  arr_delay  ... flight  tailnum origin dest  \
25        807.0             735       32.0  ...   3768   N9EAMQ    EWR  ORD   
47        931.0             940       -9.0  ...   1701   N75435    EWR  FLL   
70        852.0             832       20.0  ...   1092   N26226    LGA  ORD   
85       1011.0             941       30.0  ...   1111   N37456    EWR  MCO   
96        939.0             850       49.0  ...   3737   N508MQ    EWR  ORD   
...         ...             ...        ...  ...    ...      ...    ...  ...   
19474    1456.0            1430       26.0  ...    695   N687DL    JFK  MCO   
19498    1459.0            1459        0.0  ...   3826   N15985    EWR  ATL   
19522    1606.0            1606        0.0  ...     85   N589JB    JFK  FLL   
19550    1637.0            1641       -4.0  ...   1783   N524JB    JFK  MCO   
19551    1452.0            1355       57.0  ...   1006   N229JB    JFK  BOS   

      air_time  distance  hour  minute            time_hour velocidade média  
25       139.0       719     6       0  2013-01-01 06:00:00         5.172662  
47       151.0      1065     6      36  2013-01-01 06:00:00         7.052980  
70       135.0       733     7       0  2013-01-01 07:00:00         5.429630  
85       145.0       937     6      45  2013-01-01 06:00:00         6.462069  
96       148.0       719     7      10  2013-01-01 07:00:00         4.858108  
...        ...       ...   ...     ...                  ...              ...  
19474    142.0       944    11      30  2013-01-23 11:00:00         6.647887  
19498    121.0       746    12      30  2013-01-23 12:00:00         6.165289  
19522    156.0      1069    13       0  2013-01-23 13:00:00         6.852564  
19550    145.0       944    13      45  2013-01-23 13:00:00         6.510345  
19551     37.0       187    12      50  2013-01-23 12:00:00         5.054054  

[1004 rows x 21 columns]

In [32]:
df_med_piores['velocidade média'].mean()

6.0990621282936175

In [33]:
df_med_piores.groupby('dest').agg({'velocidade média':'mean'})

velocidade média
dest                  
ATL           6.173417
BOS           4.984730
FLL           6.781675
MCO           6.629831
ORD           5.859147

A velocidade média dos 5 destinos que mais ocorrem atrasos é de: 
ATL = 6.173417
BOS = 4.984730
FLL = 6.781675
MCO =	6.629831
ORD =	5.859147